scripts/run-pyright.py

```diff
@@ -0,0 +1,302 @@
+            env = next(
+                (
+                    env_path_spec["env"]
+                    for env_path_spec in env_path_specs
+                    if match_path(path, env_path_spec)
+                )
+            )
+        except StopIteration:
+            raise Exception(f"Could not find environment that matched path: {path}.")
+        env_path_map.setdefault(env, []).append(path)
+    return env_path_map
+
+
+def normalize_env(env: str, rebuild: bool) -> None:
+    venv_path = os.path.join(get_env_path(env), ".venv")
+    if rebuild and os.path.exists(venv_path):
+        print(f"Removing existing virtualenv for pyright environment {env}...")
+        subprocess.run(f"rm -rf {venv_path}", shell=True, check=True)
+    if not os.path.exists(venv_path):
+        print(f"Creating virtualenv for pyright environment {env}...")
+        requirements_path = f"requirements-{env}.txt"
+        shutil.copyfile(get_env_path(env, "requirements.txt"), requirements_path)
+        cmd = " && ".join(
+            [
+                f"python -m venv {venv_path}",
+                f"{venv_path}/bin/pip install -U pip setuptools wheel",
+                f"{venv_path}/bin/pip install -r {requirements_path}",
+            ]
+        )
+        subprocess.run(cmd, shell=True, check=True)
+        os.remove(requirements_path)
+    return None
+
+
+def run_pyright(env: str, paths: Sequence[str], rebuild: bool) -> RunResult:
+    normalize_env(env, rebuild)
+    with temp_pyright_config_file(env) as config_path:
+        # pipefail_cmd = "set -o pipefail"
+        pyright_cmd = " ".join(
+            ["pyright", f"--project={config_path}", "--outputjson", "--level=warning", *paths]
+        )
+        # awk_cmd = f"awk {shlex.quote(AWK_SCRIPT)} -"
+        # shell_cmd = f"{pipefail_cmd} && {pyright_cmd} | {awk_cmd}"
+        shell_cmd = pyright_cmd
+        print(f"Running pyright for environment `{env}`...")
+        print(f"  {shell_cmd}")
+        result = subprocess.run(shell_cmd, capture_output=True, shell=True, text=True)
+    return {
+        "returncode": result.returncode,
+        "output": cast(PyrightOutput, json.loads(result.stdout)),
+    }
+
+
+@contextmanager
+def temp_pyright_config_file(env: str) -> Iterator[str]:
+    with open("pyproject.toml", "r", encoding="utf-8") as f:
+        toml = tomli.loads(f.read())
+    config = toml["tool"]["pyright"]
+    config["venvPath"] = f"{PYRIGHT_ENV_ROOT}/{env}"
+    include_path = get_env_path(env, "include.txt")
+    exclude_path = get_env_path(env, "exclude.txt")
+    config["include"] = load_path_file(include_path)
+    if os.path.exists(exclude_path):
+        config["exclude"] += load_path_file(exclude_path)
+    temp_config_path = f"pyrightconfig-{env}.json"
+    print("Creating temporary pyright config file at", temp_config_path)
+    with open(temp_config_path, "w", encoding="utf-8") as f:
+        json.dump(config, f)
+    yield temp_config_path
+    os.remove(temp_config_path)
```

Comment:
seems like this should be in a finally block

Reply:
you're right